### ISSUE TITLE
Standardize commands output format for all object types

### ISSUE THREAD

**Issue by mmarchini** [nodejs/llnode#265]
tl;dr: I think we should standarize the output for all types into JSON format. The default output should hide V8 implementation details and should only show properties and elements accessible via JavaScript. A detailed mode would then expose a limited set of implementation details.

---

I would like to propose some changes to how we print objects in llnode commands. Today, every object type prints data in its own format, which can confuse new users and is harder to interpret. We also expose V8 internal states and metadata for objects by default, which is probably undesirable as the default behavior (also related: #228).

Here are some examples of how we output some object types today with `v8 inspect`:

### `JSObject`

JSObject is one of the most common type users will inspect, and it represents a JavaScript object (`{}`, `new Object`, instances of a class, etc.). Looking into a "normal" JSObject, we `<Object: ` followed by `prototypeName` and then `properties {`, which is followed by the object properties (let's call this the properties section):

```
0x2bc5c828d1:<Object: Module properties {
    .id=0x2ba8886629:<String: ".">,
    .exports=0x2bc5c82d49:<Object: Object>,
    .parent=0x2ba88822b1:<null>,
    .filename=0x2b0de4ef71:<String: "/Users/mmarchini...">,
    .loaded=0x2ba88829a1:<false>,
    .children=0x2bc5cf25a9:<Array: length=17>,
    .paths=0x2bc5cf25c9:<Array: length=17>}>
```

If the object has `elements` (numeric indexes), we'll also have a "section" for those:

```
0x2b4a71eb89:<Object: Object elements {
    [0]=0x2ba88822b1:<null>,
    [4]=0x2ba88826f1:<undefined>,
    [23]=0x2b4a71ff01:<JSRegExp source=/regexp/>,
    [25]=0x2be937a3e1:<function: c.hashmap.(anonymous function) at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/inspect-scenario.js:70:19>}
  properties {
    .some-key=<Smi: 42>,
    .other-key=0x2be9377c51:<String: "ohai">,
    .cons-string=0x2b4a71ed19:<String: "this could be a ...">,
    .internalized-string=0x2be9377d19:<String: "foobar">,
    .thin-string=0x2b4a71edc1:<String: "foobar">,
    .externalized-string=0x2be9377dd1:<String: "(external)">,
    .sliced-externalized-string=0x2b4a71ef89:<String: "(external)">,
    .array=0x2b4a71f0b9:<Array: length=6>,
    .long-array=0x2b4a71f249:<Array: length=20>,
    .array-buffer=0x2b4a71f4d1:<ArrayBuffer: backingStore=0x00000001027111f0, byteLength=5>,
    .uint8-array=0x2b4a71f561:<ArrayBufferView: backingStore=0x0000000102711830, byteOffset=0, byteLength=6>,
    .buffer=0x2b4a71f8c9:<ArrayBufferView: backingStore=0x0000000104800c00, byteOffset=2368, byteLength=6>,
    .error=0x2b4a71f981:<Object: Error>,
    .Symbol('oneSymbol')=<Smi: 42>,
    .scoped=0x2be937c7a9:<function: name at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/inspect-scenario.js:79:35>}>
```

JSObjects can also have `internal fields`, which - I think - are not directly accessible through JavaScript: 

```
0x14a8aadbb2d1:<Object: Signal
  internal fields {
    0x000014a80ee826f1}>
```

So `JSObjects` have three sections: `properties`, `elements` and `internal fields`

```
0x2bc5c84321:<Object: Object properties {
    ..js=0x2b0cbf7171:<function: Module._extensions..js at (external).js:698:37>,
    ..node=0x2b0cbf71f1:<function: Module._extensions..node at (external).js:717:39>,
    ..json=0x2b0cbf71b1:<function: Module._extensions..json at (external).js:705:39>}>
```

### `JSError`

JSErrors are essentially JSObjects, with one difference: they have a `error stack` section, with an Array of `JSFunction`s (which is a V8 implementation detail):

```
(llnode) v8 inspect 0x14a8aadb1bd9
0x14a8aadb1bd9:<Object: Error properties {
    .stack=0x14a84601f0d9:<unknown>,
    .message=0x14a8aadb1ba1:<String: "No such module: ...">,
    .Symbol()=0x14a8aadb1dc1:<Array: length=51>}
  error stack {
    0x14a8ce903c41:<function: internalBinding at internal/bootstrap/loaders.js:1:10>
    0x14a8aadb1969:<function: (anonymous) at (external).js:1:10>
    0x14a81f63a9f9:<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>
    0x14a81f63a711:<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>
    0x14a8aadac2c9:<function: (anonymous) at net.js:1:10>
    0x14a81f63a9f9:<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>
    0x14a81f63a711:<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>
    0x14a8aada0059:<function: (anonymous) at tty.js:1:10>
    0x14a81f63a9f9:<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>
    0x14a81f63a711:<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>
  }>
```

### `JSFunction`

JSFunctions don't have sections, but they are printed one level deeper than JSObjects (we also print the content of JSFunction.context, and we don't have this behavior in any other object type):

```
0x14a8aadc5609:<function: eyecatcher at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/frame-scenario.js:10:20
  context=14a8aadc5591:<Context: {
    (previous)=0x14a81f603dc1:<Context>,
    (scope_info)=0x14a824957229:<ScopeInfo: for function >,
    crasher=0x14a8aadc55c9:<function: crasher at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/frame-scenario.js:4:17>}>>
```

### `Context`

Contexts print variables present in that context mixed with implementation details enclosed in parenthesis (again, we don't have this behavior in other types):

```
0x14a8aadc5591:<Context: {
  (previous)=0x14a81f603dc1:<Context>,
  (scope_info)=0x14a824957229:<ScopeInfo: for function >,
  crasher=0x14a8aadc55c9:<function: crasher at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/frame-scenario.js:4:17>}>
```

### `JSArray`

They could follow JSObject format, but they don't have sections and we also print the length of the array:

```
0x000014a8aadefb11:<Array: length=51 {
    [0]=<Smi: 10>,
    [1]=0x14a80ee826f1:<undefined>,
    [2]=0x14a8ce903c41:<function: internalBinding at internal/bootstrap/loaders.js:1:10>,
    [3]=0x14a81f63c731:<unknown>,
    [4]=<Smi: 27>,
    [5]=<Smi: 8>,
    [6]=0x14a80ee826f1:<undefined>,
    [7]=0x14a8aadef551:<function: (anonymous) at (external).js:1:10>,
    [8]=0x14a8395accc9:<unknown>,
    [9]=<Smi: 118>,
    [10]=<Smi: 8>,
    [11]=0x14a8aaded5f1:<Object: NativeModule>,
    [12]=0x14a81f63a9f9:<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>,
    [13]=0x14a81f642281:<unknown>,
    [14]=<Smi: 301>,
    [15]=<Smi: 8>}>
```

---

This lack of standardization also makes it difficult to have good object navigation on the JS API.

We could standarize the output for all types into JSON format:
  - The default output should hide V8 implementation details and should only show properties and elements accessible via JavaScript, in a JavaScript-y format (hide memory addresses for example).
  - A detailed mode would then expose a limited set of implementation details. We could also show addresses in this detailed mode
  - Instead of addresses (which would only appear by default in `v8 findjs*` and `v8 bt` commands), we could use accessors to navigate deeper into an object. A few examples: `v8 inspect 0x2b4a71eb89.some`, `v8 inspect 0x2b4a71eb89["thin"]`, `v8 inspect 0x2b4a71eb89[4]`, `v8 inspect 0x2b4a71eb89.some[2].next.another[42]`, etc.
  - There should be a flag for the user to change the depth of the object inspection (just like `util.inspect`)

Here is an example of how the default output would look like:

```js
(llnode) v8 inspect 0x2b4a71eb89
{
  0: null,
  4: "undefined",
  23: "JSRegExp(/regexp/)",
  25: "function: c.hashmap.(anonymous function) at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/inspect-scenario.js:70:19",
  "some": 42,
  "other-key": "ohai",
  "cons": "this could be a ...",
  "internalized-string": "foobar",
  "array": {
    0: true,
    1: 1,
    2: undefined,
    3: null,
    4: "test",
    5: "function: Class at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/inspect-scenario.js:22:17"
  }
}
```

And the detailed output:

```js
(llnode) v8 inspect --detailed 0x2b4a71eb89
{
  "elements": {
    0: null,
    4: "undefined",
    23: "JSRegExp(/regexp/)",
    25: "function: c.hashmap.(anonymous function) at /Users/mmarchini/workspace/nodejs/llnode/test/fixtures/inspect-scenario.js:70:19"
  },
  "properties": {
    "some": 42,
    "other-key": "ohai",
    "array": { 
      0: true,
      1: 1,
      2: "test"
    }
  },
  "metadata": {
    "address": 0x2b4a71eb89,
    "type": "Object",
    "v8Type": "Object",
    "length": 25
  },
  "internal": {
    "internalFields": [
      "0x000014a80ee826f1"
    ],
    "stackTrace": [
      "<function: internalBinding at internal/bootstrap/loaders.js:1:10>",
      "<function: (anonymous) at (external).js:1:10>",
      "<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>"
      "<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>"
      "<function: (anonymous) at net.js:1:10>"
      "<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>"
      "<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>"
      "<function: (anonymous) at tty.js:1:10>"
      "<function: NativeModule.compile at internal/bootstrap/loaders.js:1:10>"
      "<function: NativeModule.require at internal/bootstrap/loaders.js:1:10>"
    ]
  }
}
```

If we want the address of some property of an object:

```js
(llnode) v8 inspect --detailed 0x2b4a71eb89.array
{
  "elements": {
    0: true,
    1: 1,
    2: "test"
  },
  "metadata": {
    "address": 0x2b4a71f0b9,
    "type": "Array",
    "v8Type": "JSArray",
    "length": 3
  }
}
```

There should also be a way to navigate the `internal` properties (such as `stackTrace`), but we can still come up with a syntax for that.

Please let me know what you think. I'll start working on a proof-of-concept for this proposal.
